### ISSUE TITLE
PHP CLI segmentation faults with datadog-php-tracer 0.29.0

### ISSUE THREAD

**Issue by ctrlrsf** [DataDog/dd-trace-php#507]
We recently noticed intermittent PHP seg faults for CLI scripts. Although we've seen it with all PHP versions we've tested (7.0 from Ubuntu 16.04, 7.2, and 7.3), below is simplest way we've found to reproduce it.

```
# Seg faults while just starting php and calling sleep(.001)
$ while true; do php -r 'sleep(.001);'; /bin/echo -ne '.';  done
..................................Segmentation fault (core dumped)
................................................................................................Segmentation fault (core dumped)
.............................................................................................................................................................Segmentation fault (core dumped)
.....................................................................................................................................................................................................Segmentation fault (core dumped)
.....Segmentation fault (core dumped)
.......Segmentation fault (core dumped)
..................Segmentation fault (core dumped)
................Segmentation fault (core dumped)
.....Segmentation fault (core dumped)
...........................Segmentation fault (core dumped)
............Segmentation fault (core dumped)
............^CSegmentation fault (core dumped)
.............^C

# Uninstall datadog-php-tracer
$ dpkg --purge datadog-php-tracer
(Reading database ... 59348 files and directories currently installed.)
Removing datadog-php-tracer (0.29.0) ...
Purging configuration files for datadog-php-tracer (0.29.0) ...
dpkg: warning: while removing datadog-php-tracer, directory '/opt/datadog-php/etc' not empty so not removed
$ find /etc/php -name '*ddtrace*' -delete -print
/etc/php/7.0/cli/conf.d/99-ddtrace-custom.ini
/etc/php/7.0/cli/conf.d/98-ddtrace.ini

# No more seg faults
$ while true; do php -r 'sleep(.001);'; /bin/echo -ne '.';  done
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
..........................................................................................................................
......................^C
```

Versions:
```
$ php -v
PHP 7.0.33-0ubuntu0.16.04.5 (cli) ( NTS )
Copyright (c) 1997-2017 The PHP Group
Zend Engine v3.0.0, Copyright (c) 1998-2017 Zend Technologies
    with Zend OPcache v7.0.33-0ubuntu0.16.04.5, Copyright (c) 1999-2017, by Zend Technologies

$ uname -a
Linux xyz 4.4.0-154-generic #181-Ubuntu SMP Tue Jun 25 05:29:03 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.6 LTS
Release:	16.04
Codename:	xenial

$ dpkg -l | grep datadog-php
ii  datadog-php-tracer               0.29.0                                                   amd64        no description given
```

Here's a history of the seg faults while testing:
```
[ 7125.066568] php[12457]: segfault at 7fe17be130fd ip 00007fe17be130fd sp 00007fe176f83f18 error 14 in libbsd.so.0.8.2[7fe17cc57000+13000]
[ 7160.110386] php[13862]: segfault at 7f2e6b6130fd ip 00007f2e6b6130fd sp 00007f2e66783f18 error 14 in librt-2.23.so[7f2e6e46e000+7000]
[ 7190.349904] php[15916]: segfault at 7fe6ce4130fd ip 00007fe6ce4130fd sp 00007fe6c9583f18 error 14 in tokenizer.so[7fe6ce62c000+4000]
[ 7190.453141] php[15921]: segfault at 7f72aec130fd ip 00007f72aec130fd sp 00007f72a9d83f18 error 14 in tokenizer.so[7f72aee2c000+4000]
[ 7191.644470] php[16060]: segfault at 7f05ec6130fd ip 00007f05ec6130fd sp 00007f05e7783f18 error 14 in librt-2.23.so[7f05ef46e000+7000]
[ 7191.868619] php[16069]: segfault at 7efd896130fd ip 00007efd896130fd sp 00007efd84783f18 error 14 in librt-2.23.so[7efd8c46e000+7000]
[ 7192.966512] php[16159]: segfault at 7f5a24a130fd ip 00007f5a24a130fd sp 00007f5a1fb83f18 error 14 in librt-2.23.so[7f5a2786e000+7000]
[ 7193.344117] php[16179]: segfault at 7f84ac6130fd ip 00007f84ac6130fd sp 00007f84a7783f18 error 14 in ftp.so[7f84ae73d000+c000]
[ 7194.357279] php[16207]: segfault at 7f1c77c130fd ip 00007f1c77c130fd sp 00007f1c72d83f18 error 14 in librt-2.23.so[7f1c7aa6e000+7000]
[ 7195.106510] php[16283]: segfault at 7fc6f1a130fd ip 00007fc6f1a130fd sp 00007fc6ecb83f18 error 14 in librt-2.23.so[7fc6f486e000+7000]
[ 7195.351185] php[16292]: segfault at 7f27b6e130fd ip 00007f27b6e130fd sp 00007f27b1f83f18 error 14 in librt-2.23.so[7f27b9c6e000+7000]
[ 7196.207782] php[16300]: segfault at 7f5bb46130fd ip 00007f5bb46130fd sp 00007f5baf783f18 error 14 in tokenizer.so[7f5bb482c000+4000]
[ 7196.442759] php[16309]: segfault at 7f523cc130fd ip 00007f523cc130fd sp 00007f5237d83f18 error 14 in libbsd.so.0.8.2[7f523da57000+13000]
[ 7197.448663] php[16336]: segfault at 7f9cf76130fd ip 00007f9cf76130fd sp 00007f9cf2783f18 error 14 in tokenizer.so[7f9cf782c000+4000]
[ 7198.119549] php[16363]: segfault at 7f54bb0130fd ip 00007f54bb0130fd sp 00007f54b6183f18 error 14 in tokenizer.so[7f54bb22c000+4000]
[ 7198.508779] php[16371]: segfault at 7f56362130fd ip 00007f56362130fd sp 00007f5631383f18 error 14 in tokenizer.so[7f563642c000+4000]
[ 7199.346005] php[16379]: segfault at 7fdfc3a130fd ip 00007fdfc3a130fd sp 00007fdfbeb83f18 error 14 in libbsd.so.0.8.2[7fdfc4857000+13000]
[ 7200.741057] php[16431]: segfault at 7fb6d16130fd ip 00007fb6d16130fd sp 00007fb6cc783f18 error 14 in tokenizer.so[7fb6d182c000+4000]
[ 7203.704685] php[16736]: segfault at 7f5edd4130fd ip 00007f5edd4130fd sp 00007f5ed8583f18 error 14 in tokenizer.so[7f5edd62c000+4000]
[ 7211.745471] php[17829]: segfault at 7f7215a130fd ip 00007f7215a130fd sp 00007f7210b83f18 error 14 in librt-2.23.so[7f721886e000+7000]
[ 7221.579679] php[19206]: segfault at 7fd674e130fd ip 00007fd674e130fd sp 00007fd66ff83f18 error 14 in tokenizer.so[7fd67502c000+4000]
[ 7222.094001] php[19268]: segfault at 7f11a86130fd ip 00007f11a86130fd sp 00007f11a3783f18 error 14 in librt-2.23.so[7f11ab46e000+7000]
[ 7224.004829] php[19618]: segfault at 7f762f4130fd ip 00007f762f4130fd sp 00007f762a583f18 error 14 in librt-2.23.so[7f763226e000+7000]
[ 7224.673932] php[19710]: segfault at 7ff6d4e130fd ip 00007ff6d4e130fd sp 00007ff6cff83f18 error 14 in librt-2.23.so[7ff6d7c6e000+7000]
[ 7225.062878] php[19812]: segfault at 7fa9044130fd ip 00007fa9044130fd sp 00007fa8ff583f18 error 14 in librt-2.23.so[7fa90726e000+7000]
[ 7225.783884] php[19908]: segfault at 7f19ea6130fd ip 00007f19ea6130fd sp 00007f19e5783f18 error 14 in librt-2.23.so[7f19ed46e000+7000]
[ 7226.166872] php[19948]: segfault at 7f7fca2130fd ip 00007f7fca2130fd sp 00007f7fc5383f18 error 14 in librt-2.23.so[7f7fcd06e000+7000]
[ 7226.473335] php[19977]: segfault at 7fd7e86130fd ip 00007fd7e86130fd sp 00007fd7e3783f18 error 14 in librt-2.23.so[7fd7eb46e000+7000]
[ 7228.906859] php[20344]: segfault at 7fa2fb0130fd ip 00007fa2fb0130fd sp 00007fa2f6183f18 error 14 in tokenizer.so[7fa2fb22c000+4000]
[ 7230.304162] php[20541]: segfault at 7f0e4c8130fd ip 00007f0e4c8130fd sp 00007f0e47983f18 error 14 in librt-2.23.so[7f0e4f66e000+7000]
[ 7231.026733] php[20637]: segfault at 7fa82aa130fd ip 00007fa82aa130fd sp 00007fa825b83f18 error 14 in librt-2.23.so[7fa82d86e000+7000]
[ 7231.394058] php[20674]: segfault at 7fda584130fd ip 00007fda584130fd sp 00007fda53583f18 error 14 in librt-2.23.so[7fda5b26e000+7000]
[ 7231.927138] php[20737]: segfault at 7f604ce130fd ip 00007f604ce130fd sp 00007f6047f83f18 error 14 in librt-2.23.so[7f604fc6e000+7000]
[ 7232.364040] php[20785]: segfault at 7f3fa98130fd ip 00007f3fa98130fd sp 00007f3fa4983f18 error 14 in librt-2.23.so[7f3fac66e000+7000]
[ 7232.996243] php[20866]: segfault at 7efd908130fd ip 00007efd908130fd sp 00007efd8b983f18 error 14 in librt-2.23.so[7efd9366e000+7000]
[ 7233.068116] php[20874]: segfault at 7fe45da130fd ip 00007fe45da130fd sp 00007fe458b83f18 error 14 in librt-2.23.so[7fe46086e000+7000]
[ 7234.374241] php[21002]: segfault at 7f47ff2130fd ip 00007f47ff2130fd sp 00007f47fa383f18 error 14 in librt-2.23.so[7f480206e000+7000]
```

**Comment by SammyK** [comment]
Thanks so much for all the detailed information @ctrlrsf! We've been able to get an environment similar to yours set up and are in the process of trying to reproduce the segfault. We will let you know as soon as we have more information. :)

**Comment by pawelchcki** [comment]
@ctrlrsf Thanks for reporting!
This issue was caused by race condition in our module shutdown code, where sometimes our PHP module was shutdown faster than background thread, causing the segmentation faults.

The fix is already in the master. We're planning to release a version containing that fix next week.
Will update this issue when thats done

**Comment by ctrlrsf** [comment]
Great to hear and thanks for your help as always.

**Comment by labbati** [comment]
Hi @ctrlrsf and thanks for your patience. Did you have a change to test the [0.30.0](https://github.com/DataDog/dd-trace-php/releases/tag/0.30.0) release to confirm that resolved your issue?

**Comment by morrisonlevi** [comment]
Hey, @ctrlrsf, just following up. Have you been able to confirm that an 0.30 release such as [0.30.2](https://github.com/DataDog/dd-trace-php/releases/tag/0.30.2) fixes the issue for you?

**Comment by labbati** [comment]
Hi @ctrlrsf, as mentioned by @pawelchcki this issue has been resolved and no longer exists after 0.30.2. I am closing it, please feel free to re-open if you experience this again.
